### PR TITLE
16 - Created the archive functionality for projects

### LINKED ISSUE #16: Create a controller to archive a project

### LINKED ISSUE #34: [BUG] - Archive project button
**Describe the bug**
The archive project button in the professor's view doesn't work

**To Reproduce**
1. Go to /professor
2. Click on archive button of one of the projects

**Expected behavior**
It is expected to show a modal asking for confirmation of the deletion request
It is expected to then archive the project

**Actual behavior**
Nothing actually happens when you click the edit button

**Screenshots**

**Additional context**

### PR DESCRIPTION
**Pull request by blessingomotayo**
## Proposed changes

Added the archive functionality of a project in the professor view

## Types of changes

What types of changes does your code introduce?
_Put an `x` in the boxes that apply_

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. This is simply a reminder of what we are going to look for before merging your code._
- [x] The title is in the format [Issue number]-[Issue Title]
- [x] Unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [x] I have moved issue to the “Need review” columns of projects

## Tasks completed
- When the archive button is clicked on a project, it is removed from the list of projects the professor sees
- Can go to /archivedprojects to see the projects that are archived

## Tests performed
- Tests that the Archived Projects page loads correctly
- Tests that when a project is added, and call archiveProject(), it should query the projects where archivedProject is not null and confirm.

Closes #34 
Closes #16 

## Screenshots (if applicable)
Include screenshots of the front-end changes that were made

## Must Do Before Approving
_Put an `x` in the boxes that apply.
- [x] Review the code changes and check for code smells and or make comments if needed

Put an `x` in the box below if you need the reviewer to double check the front-end changes you made
- [x] Tophat Needed (Steps: checkout the branch and run the application locally to determine if the front-end change looks the same as in this PR)

### CODE REVIEW

**Comment by blessingomotayo** [comment]
The "ArchivedProjects" html is only created for us to see where all the archived projects went

**Comment by arungalva** [comment]
image

File: src/test/java/com/eggplant/emoji/app/JPAProjectTests.java [hunk 1]
```diff
@@ -147,4 +149,84 @@ public void getProjects() throws Exception {
...
+        assertEquals("Test Project", projectExp.getProjectName());
+        assertEquals("Test Project Description", projectExp.getDescription());
+        assertEquals(2, projectExp.getMinNumberOfStudents());
+        assertEquals(5, projectExp.getMaxNumberOfStudents());
+        assertEquals(null, projectExp.getArchivedDate());
+
+        EntityManagerFactory emf = Persistence.createEntityManagerFactory("jpa-test");
+        EntityManager em = emf.createEntityManager();
+        EntityTransaction tx = em.getTransaction();
+        Query q = em.createQuery("SELECT p FROM Project p WHERE p.projectName = 'Test Project'");
```
Comment (arungalva) [inline src/test/java/com/eggplant/emoji/app/JPAProjectTests.java]: We want to move away from entitymanager. Could you please refactor this to use repository?

File: src/test/java/com/eggplant/emoji/app/JPAProjectTests.java [hunk 2]
```diff
@@ -147,4 +149,84 @@ public void getProjects() throws Exception {
...
+
+        Project p = results.get(0);
+        p.archiveProject();
+        assertNotNull(p.getArchivedDate());
+
+        tx.begin();
+        em.persist(p);
+        tx.commit();
+
+        Query query = em.createQuery("SELECT p FROM Project p WHERE p.archivedDate != null");
```
Comment (arungalva) [inline src/test/java/com/eggplant/emoji/app/JPAProjectTests.java]: Would you be able to refactor this to use repositories instead of entity manager?

**Comment by blessingomotayo** [comment]
Yeah I had to do that too @arungalva to run Shasthra's changes from master. Not sure if it's SQLite because he also had to do that too. But from our conversation yesterday, when the app is deployed, it does create a new sample.db, if I'm not mistaken. @ShasRanasinghe

**Comment by ShasRanasinghe** [comment]
Yeah, that's what should be happening.

**Comment by ShasRanasinghe** [comment]
@blessingomotayo When archiving things, I always thought the way to go about it was to have a separate table just for archived projects. Is there a benefit to the archive date way of doing it?

**Comment by blessingomotayo** [comment]
Michael and I talked about a way to do it and I think having an attribute instead of a new table is better and maybe faster. If we have a different table, the projects seem to be duplicated to me and the only thing that is different is that they are archived. So, I think leaving them in the same table and just adding one attribute to determine if it is archived or not is less of a hassle. 

What do you think?

**Comment by arungalva** [comment]
I like the way blessing did it. However, I would have had a column called state which would either be ACTIVE or ARCHIVED to track the state of the project but the way you did it now works too. I think in the future we should add the state column though

**Comment by ShasRanasinghe** [comment]
Hmm, ok yeah that makes sense actually.
thanks @blessingomotayo

**Comment by blessingomotayo** [comment]
Yep I see what you mean @arungalva . We can explore that option in the future. 👍

**Comment by arungalva** [comment]
Could you create a feature request to be able to unarchive projects?

**Comment by arungalva** [comment]
Since you are using entity manager, when #70 gets merged, your code wouldn't work anymore. I think we should get #70 merged first. You could pull those changes and rewrite the archive controller to use repository instead of entity manager

**Comment by blessingomotayo** [comment]
Yeah, working on that already

**Pull request merged**